In [13]:
import pandas as pd
import statsmodels.api as sm
df = pd.read_csv('C:\\Users\\divit\\Desktop\\Linear Regression ( Program + Use case )\\Advertising.csv')
df.head()

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [15]:
df.drop(['Unnamed: 0'], axis = 1,inplace = True)
df.head()

# The column shows of expenditure( the amount of money that is spent ) of every headers

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

y = m1x1 + m2x2 + m3x3,
y = best fit line, m1 = age, m2 = Experience, m3 = salary. c= intercept

Multi collinearity says that if your independent features are correlated with more than 90% we have to now resolve this issue

In [8]:
# independent features
x = df[['TV',"Radio", 'Newspaper']]

# dependent features
y = df[['Sales']]

ols is ordinary least square 
It is used to predict values of a continuous response variable using one or more explanatory variables and can also identify the strength of the relationships between these variables.

ols will make a seprate column of our intercept C, it will give the value 1 to it & our c will be the coefficient of 1. so it will be like ( c * 1 )

In [17]:
# adding a constant x
x = sm.add_constant(x)

In [18]:
x

const     TV  Radio  Newspaper
0      1.0  230.1   37.8       69.2
1      1.0   44.5   39.3       45.1
2      1.0   17.2   45.9       69.3
3      1.0  151.5   41.3       58.5
4      1.0  180.8   10.8       58.4
..     ...    ...    ...        ...
195    1.0   38.2    3.7       13.8
196    1.0   94.2    4.9        8.1
197    1.0  177.0    9.3        6.4
198    1.0  283.6   42.0       66.2
199    1.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [19]:
# fit a OLS model with intercept TV & Radio
model = sm.OLS(y,x).fit()
model

In [21]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sun, 25 Jul 2021   Prob (F-statistic):           1.58e-96
Time:                        15:51:25   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The above graph will make us understand if there is any multi collinearility between these features.
Eg. if TV & Radio have high correlation ( or ) Radio or Newspaper have high correlation 

The constant in the summary is our C ( intercept )
in tv, radio you see somes values that means that you need to increase the expenditure on these features by those values (0.458, 0.1885) 
The newspaper has negative value, it means that we do not need to do that much expenditure

The r squared value & adjusted R squared values are perfect, 
The standard error values are low ( this indicates that none of the features are having multicollinearity issues because p values are also low but p value i high for newspaper because the newspaper shows that we are making unnecessary expenditure on the newspapers so we have to reduce it)
If there is some relation beween the independent features the standard errors will be high 

In [27]:
import matplotlib.pyplot as pt
x.iloc[:,1:].corr()

# here we can see that there isn't much correlation b/w the featurez

TV     Radio  Newspaper
TV         1.000000  0.054809   0.056648
Radio      0.054809  1.000000   0.354104
Newspaper  0.056648  0.354104   1.000000

In [34]:
import pandas as pd
import statsmodels.api as sm
df1 = pd.read_csv('C:\\Users\\divit\\Desktop\\Linear Regression ( Program + Use case )\\Salary_Data1.csv')
df.head()

YearsExperience   Age  Salary
0              1.1  21.0   39343
1              1.3  21.5   46205
2              1.5  21.7   37731
3              2.0  22.0   43525
4              2.2  22.2   39891

In [46]:
x1 = df1[['YearsExperience', 'Age']]
y1 = df1['Salary']

In [63]:
x1 = sm.add_constant(x1)

In [64]:
x1.head()

const  YearsExperience   Age
0    1.0              1.1  21.0
1    1.0              1.3  21.5
2    1.0              1.5  21.7
3    1.0              2.0  22.0
4    1.0              2.2  22.2

In [65]:
modell  = sm.OLS(x1, y1).fit()
modell

In [70]:
modell.summary()

ValueError: shapes (30,3) and (30,3) not aligned: 3 (dim 1) != 30 (dim 0)